blockchain/validator_set.go

```diff
@@ -0,0 +1,273 @@
+package blockchain
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+
+	acm "github.com/hyperledger/burrow/account"
+	rpc_core "github.com/tendermint/tendermint/rpc/core"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+	tm_types "github.com/tendermint/tendermint/types"
+)
+
+type ValidatorSet interface {
+	TotalPower() int
+	SetMaximumPower(maximumPower int)
+	MaximumPower() int
+	AdjustPower(height int64) error
+	Validators() []acm.Validator
+	SetLeavers() []acm.Validator
+	JoinToTheSet(validator acm.Validator) error
+	LeaveFromTheSet(validator acm.Validator) error
+	IsValidatorInSet(address acm.Address) bool
+	JSONBytes() ([]byte, error)
+}
+
+type validatorListProxy interface {
+	Validators(height int64) (*ctypes.ResultValidators, error)
+}
+
+type _validatorListProxy struct{}
+
+func (vlp _validatorListProxy) Validators(height int64) (*ctypes.ResultValidators, error) {
+	result, err := rpc_core.Validators(&height)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+type validatorSet struct {
+	maximumPower int
+	validators   []acm.Validator
+	setLeavers   []acm.Validator
+	proxy        validatorListProxy
+}
+
+func newValidatorSet(maximumPower int, validators []acm.Validator) *validatorSet {
+	set := &validatorSet{
+		validators: validators,
+		proxy:      _validatorListProxy{},
+	}
+	set.SetMaximumPower(maximumPower)
+	return set
+}
+
+func (vs *validatorSet) TotalPower() int {
+	return len(vs.validators)
+}
+
+func (vs *validatorSet) SetMaximumPower(maximumPower int) {
+	if maximumPower > 90 {
+		maximumPower = 90
+	}
+
+	if maximumPower < 4 {
+		maximumPower = 4
+	}
+
+	vs.maximumPower = maximumPower
+}
+
+func (vs *validatorSet) MaximumPower() int {
+	return vs.maximumPower
+}
+
+func (vs *validatorSet) AdjustPower(height int64) error {
+	/// empty slice
+	vs.setLeavers = vs.setLeavers[:0]
+
+	dif := vs.TotalPower() - vs.maximumPower
+	if dif <= 0 {
+		return nil
+	}
+
+	limit := int(math.Floor(float64(vs.maximumPower*1/3))) - 1
```

Comment:
Currently is not doing any float arithmetic, you mean `limit := int(math.Floor(float64(vs.maximumPower)/3)) - 1` but that itself can be simplified to just: `limit := vs.MaximumPower/3 - 1`, and it should be since in float arithmetic is not always deterministic.


However I am not convinced this is the value you mean, I think you want `limit := int(math.Ceil(float64(vs.maximumPower)/3)) - 1` since you want to limit `dif` to strictly less than 1/3 of maximum power (I think). So for example 10/3 = 3.33... so 3 is acceptable (< 1/3), whereas your formula would set it to 2 (floor(3.33) -1). Taking the ceiling would give for max power 10: ceil(10/3) - 1 = 3, for max power 9: ceil(9/3) -1 = 2 (which is what we want for strictly less than.

You can compute the ceiling with pure integer arithemtic with `limit := (vs.maximumPower + 2)/3` (the general integer formula is ceil(float64(n)/float64(d)) == (n + d - 1)/d for integers n,d (disregarding overflow))

Reply:
Note you still have an error in the code `float64(vs.maximumPower*1/3)` does nothing because `vs.maximumPower*1/3` has already be integer divided.

You should just use `vs.maxmimPower/3 - 1` - it's simpler, faster, and more correct.